### ISSUE TITLE
setSecurityGroupName() not working as expected.

### ISSUE THREAD

**Issue by cwitte4191** [hazelcast/hazelcast-aws#94]
TLDR: setSecurityGroupName() completely prevents the cluster from identifying eligible peers.

With several ec2 instances running on the same region (but in different security groups), the cluster correctly discovers itself within the limitations of the VPC(s).  However, with this configuration, each ec2 instance attempts to connect to instances in foreign security groups that it can't (and shouldn't) access.  Unfortunately, the setSecurityGroupName() prevents the cluster from identifying any peers at all (instead of limiting the discovery to the securityGroup).
I suspect the issue may be that the filter only works with ec2 classic (quote is from the aws cli docs, included below). 

> group-name - The name of the security group for the instance. EC2-Classic only.

If this is a known issue, it would be nice to note it.   If it is not an issue, I have to idea why it didn't work.

https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-instances.html

**Comment by mesutcelik** [comment]
Hi @cwitte4191 ,
I didn't know such an issue. 

What is the main reason you set the security group name although you have multiple security groups in your VPC? 
Can you please also share your hazelcast configuration and version of hazelcast?

As you described, CLI documentation mentions this as EC2-Classic only but hazelcast-aws does not use CLI but query directly the REST API.

Any more info would be appreciated!

**Comment by cwitte4191** [comment]
I tried filtering on tags instead of security group, and that isn't working right now either.  I'll troubleshoot it a little more to see what i can turn up.

**Comment by leszko** [comment]
@cwitte4191 , tags should work, I checked it some time ago. Could you share your AWS EC2 instances tags and your Hazelcast configuration? Then, I could help.

**Comment by feanor07** [comment]
@cwitte4191 seems right about `security-group-name` filter not working for EC2s within within a VPC (non EC2-Classic instances). Check out the following dummy application and corresponding documentation illustrating the case https://github.com/feanor07/aws-describe-instances

I will try to create another project with `hazelcast` in action; but if I am not mistaken this proves @cwitte4191 's claim.

**Comment by feanor07** [comment]
please ignore my previous comment as it has nothing to do with how `hazelcast-aws` works indeed. just understood what @mesutcelik mentioned at https://github.com/hazelcast/hazelcast-aws/issues/94#issuecomment-484835201

anyway I did some testing today and created the following repo https://github.com/feanor07/hazelcast-aws-discovery
in that repo I used the dependency `hazelcast-all` as suggested in documentation and deployed to `aws elasticbeanstalk`.

I deployed 2 separate environments with each containing 2 instances. In the first environment I tested `security-group-name` filtering and in the second `tag:key-value` filtering. Both worked fine!

Then I tested using my old dependency versions which were causing issues:

```
		<dependency>
			<groupId>com.hazelcast</groupId>
			<artifactId>hazelcast</artifactId>
			<version>3.12.1</version>
		</dependency>
		<dependency>
			<groupId>com.hazelcast</groupId>
			<artifactId>hazelcast-spring</artifactId>
			<version>3.12.1</version>
		</dependency>
		<dependency>
			<groupId>com.hazelcast</groupId>
			<artifactId>hazelcast-aws</artifactId>
			<version>2.4</version>
		</dependency>
```

When I changed dependencies like this and re-deploy the jar file upon re-generating it, none of the environments worked. The instances could not find each other. Sample warning from logs: `2020-07-21 21:05:56.080  WARN 7624 --- [cached.thread-3] com.hazelcast.aws.AwsDiscoveryStrategy   : No EC2 instances found!`

Then I changed the dependencies to:

```
		<dependency>
			<groupId>com.hazelcast</groupId>
			<artifactId>hazelcast</artifactId>
			<version>4.0.2</version>
		</dependency>
		<dependency>
			<groupId>com.hazelcast</groupId>
			<artifactId>hazelcast-spring</artifactId>
			<version>4.0.2</version>
		</dependency>
		<dependency>
			<groupId>com.hazelcast</groupId>
			<artifactId>hazelcast-aws</artifactId>
			<version>3.2</version>
		</dependency>
```

and both environments worked as expected again.

To sum up, I am not sure what the issue is with the failing versions I had tested. I hope this helps.

**Comment by leszko** [comment]
Thanks for the explanation @feanor07 

Closing this issue as it seems there is no problem with the current version.
